File: programmers/graph/Lv2_타겟넘버.swift
//
//  Lv2_타겟넘버.swift
//  Algorithm_Swift
//
//  Created by Johyeon Yoon on 2021/08/04.
//

// 타겟넘버
// dfs로 풀이

import Foundation

func dfs(_ number: [Int], _ depth: Int, _ target: Int, _ value: Int, answer: inout Int) {

    if depth >= number.count {
        if target == value { answer = answer + 1 }
        return
    }

    dfs(number, depth + 1, target, value + number[depth], answer: &answer)
    dfs(number, depth + 1, target, value - number[depth], answer: &answer)

}

func solution(_ numbers:[Int], _ target:Int) -> Int {

    var answer = 0
    dfs(numbers, 0, target, 0, answer: &answer)

    return answer
}


print(solution([1, 1, 1, 1, 1], 3)) //5

File: programmers/graph/Lv2_타겟넘버.cpp
//
//  Lv2_타겟넘버.cpp
//  AlgorithmCpp
//
//  Created by Johyeon Yoon on 2021/10/22.
//

#include <string>
#include <vector>
#include <iostream>
using namespace std;


void dfs(vector<int> &numbers, int depth, int curr_sum, int &target, int &answer) {
    if (depth >= numbers.size()) {
        if (curr_sum == target) {
            answer ++;
        }
        return;
    }
    dfs(numbers, depth + 1, curr_sum + numbers[depth], target, answer);
    dfs(numbers, depth + 1, curr_sum - numbers[depth], target, answer);
}

int solution(vector<int> numbers, int target) {
    int answer = 0;

    dfs(numbers, 0, 0, target, answer);
    return answer;
}
